sympy/sets/ordinals.py

```diff
@@ -0,0 +1,190 @@
+import operator
+
+class OrdinalPow(Basic):
+    """
+    Represents ordinal exponential and multiplication terms one of the
+    building blocks of the Ordinal class.
+    In OrdinalPow(a, b) a represents exponent and b represents multiplicity.
+    """
+    def __new__(cls, *terms):
+        if not all (isinstance(i, (int, Ordinal)) for i in terms):
+            raise TypeError("arguments must be an integer or an Ordinal instance")
+        if not len(terms) == 2:
+            raise TypeError("number of arguments is invalid")
+        return Basic.__new__(cls, *terms)
+
+    @property
+    def exp(self):
+        return self.args[0]
+
+    @property
+    def mult(self):
+        return self.args[1]
+
+    @property
+    def contains_ordinal(self):
+        return any(isinstance(x, Ordinal) for x in self.args)
+
+    def _compare_term(self, other, op):
+        if self.exp == other.exp:
+            return op(self.mult, other.mult)
+        else:
+            return op(self.exp, other.exp)
+
+
+
+class Ordinal(Basic):
+    """
+    Represents ordinals in Cantor normal form.
+
+    Internally, this class is just a list of instances of OrdinalPow
+    Examples
+    ========
+    >>> from sympy.sets import Ordinal, Ordinals, OrdinalPow
+    >>> Ordinals.w  # doctest: +SKIP
+    w
+    >>> a = Ordinals.w
+    >>> a.is_limit_ordinal
+    True
+    >>> Ordinal([OrdinalPow(3,2)])  # doctest: +SKIP
+    {w**3}*2
+    >>> Ordinal([OrdinalPow(5,1),OrdinalPow(3,2)])  # doctest: +SKIP
+    w**5 + {w**3}*2
+    """
+    def __new__(cls, terms):
+        obj = super(Ordinal, cls).__new__(cls, *terms)
+        powers = [i.exp for i in obj.args]
+        if not all(earlier >= later for earlier, later in zip(powers, powers[1:])):
+            raise ValueError("powers must be in decreasing order")
+        return obj
+
+    @property
+    def is_successor_ordinal(self):
+        return (self.args[-1].exp == 0)
+
+    @property
+    def is_limit_ordinal(self):
+        return not self.is_successor_ordinal
+
+    def _insert_index(self, a, x, lo=0, hi=None):
+        """Return the index where to insert item x in list a,
+        assuming a is sorted in descending order
+
+        Essentially, the function returns number of elements in a which are >= than x.
+        """
+        if lo < 0:
+            raise ValueError('lo must be non-negative')
+        if hi is None:
+            hi = len(a)
+        while lo < hi:
+            mid = (lo+hi)//2
+            if x > a[mid]: hi = mid
+            else: lo = mid+1
+        return lo
+
+    def _convert_integer(self, integer_value):
+        return Ordinal([OrdinalPow(0, integer_value)])
+
+    def __lt__(self, other):
+        if self == other:
+            return False
+
+        elif isinstance(other, int):
+            return operator.lt(self, self._convert_integer(other))
+
+        elif isinstance(other, Ordinal):
+            index =  map(operator.eq, self.args, other.args).index(False)
```

Comment:
I think that a sequential comparison like this would be more readable:

    for ts, to in zip(self.args, other.args):
        if ts != to:
            return ts < to
    return len(self.args) < len(other.args)

I would also implement `__lt__`  for the terms.